core/types/transaction.go

```diff
@@ -350,31 +420,44 @@ func (tx *Transaction) Size() common.StorageSize {
 	return common.StorageSize(c)
 }
 
+func (tx *Transaction) EffectiveGasPrice(dst *big.Int, baseFee *big.Int) *big.Int {
+	return tx.inner.effectiveGasPrice(dst, baseFee)
+}
+
 // AsMessage returns the transaction as a core.Message.
-func (tx *Transaction) AsMessage(s Signer, balanceFee *big.Int, number *big.Int) (Message, error) {
+func (tx *Transaction) AsMessage(s Signer, balanceFee *big.Int, header *Header) (Message, error) {
 	msg := Message{
 		nonce:           tx.Nonce(),
 		gasLimit:        tx.Gas(),
 		gasPrice:        new(big.Int).Set(tx.GasPrice()),
+		gasFeeCap:       new(big.Int).Set(tx.GasFeeCap()),
+		gasTipCap:       new(big.Int).Set(tx.GasTipCap()),
 		to:              tx.To(),
 		amount:          tx.Value(),
 		data:            tx.Data(),
 		accessList:      tx.AccessList(),
-		checkNonce:      true,
+		isFake:          false,
 		balanceTokenFee: balanceFee,
 	}
 
-	var err error
-	msg.from, err = Sender(s, tx)
+	// If baseFee provided, set gasPrice to effectiveGasPrice.
+	if header != nil && header.BaseFee != nil {
+		msg.gasPrice = math.BigMin(msg.gasPrice.Add(msg.gasTipCap, header.BaseFee), msg.gasFeeCap)
```

Comment:
This gasPrice will get overwritten if balanceFee != nil and other condition, is it expected?

Reply:
The code is improved.